File: NewsAppWIithLoader/app/src/main/java/com/example/newsappwiithloader/MainActivity.java
package com.example.newsappwiithloader;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.SearchView;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.Loader;

import android.content.Intent;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.Uri;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.Button;
import android.widget.ListView;
import android.widget.TextView;
import android.widget.Toast;

import java.util.List;

public class MainActivity extends AppCompatActivity implements LoaderManager.LoaderCallbacks<List<News>> {
    AlertDialog dialog;
    ListView listView;
    List<News> news;
    SearchView searchView;
    TextView emptyTextView;
    Button searchButton;
    String baseUrl =  "https://content.guardianapis.com/search"; // https://content.guardianapis.com/search?show-tags=contributor&q=apple&q=apple&api-key=d14c82f4-c6be-45fd-a246-6c8fca5fc726       after ? will be the addition information
    String apiKey = "d14c82f4-c6be-45fd-a246-6c8fca5fc726";
    final int LOADER_ID_NEWS = 1;
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        listView = findViewById(R.id.main_listView);
        searchView = findViewById(R.id.main_searchView);
        searchButton = findViewById(R.id.main_search_button);
        emptyTextView = findViewById(R.id.main_empty_news_textView);
        dialog = getDialog();

        searchButton.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                if(isNetWorkConnected()){
                    dialog.show();
                    //getSupportLoaderManager().initLoader(LOADER_ID_NEWS, null, MainActivity.this);
                    getSupportLoaderManager().restartLoader(LOADER_ID_NEWS, null, MainActivity.this);
                }else {
                    Toast.makeText(getBaseContext(),"Need to connect to the Internet",Toast.LENGTH_SHORT).show();
                }
            }
        });
        listView.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int position, long l) {
                Uri uri = Uri.parse(news.get(position).getWeb());
                Intent intent = new Intent(Intent.ACTION_VIEW,uri);
                startActivity(intent);
            }
        });
    }

    private String getCompleteUrl (){
        Uri uri = Uri.parse(baseUrl);
        Uri.Builder builder = uri.buildUpon();

        builder.appendQueryParameter("q", searchView.getQuery().toString());
        builder.appendQueryParameter("show-tags", "contributor");
        builder.appendQueryParameter("api-key",apiKey);

        Log.v("UrlShow",builder.toString());

        return builder.toString();
    }

    private boolean isNetWorkConnected(){
        ConnectivityManager connectivityManager = (ConnectivityManager) getSystemService(CONNECTIVITY_SERVICE);
        NetworkInfo networkInfo = connectivityManager.getActiveNetworkInfo();
        return networkInfo!=null && networkInfo.isConnected();
    }

    private AlertDialog getDialog(){
        AlertDialog.Builder builder = new AlertDialog.Builder(MainActivity.this);
        builder.setMessage("loading");
        builder.setCancelable(false); // cannot close through click other buttons
        dialog = builder.create();
        return dialog;
    }

    @NonNull
    @Override
    public Loader<List<News>> onCreateLoader(int id, @Nullable Bundle args) { //check by id to see did we run that, so initial loader will only run once.
        return new NewsLoader(this,getCompleteUrl());
    }

    @Override
    public void onLoadFinished(@NonNull Loader<List<News>> loader, List<News> data) {
        dialog.dismiss();
        if(data.isEmpty()){
            emptyTextView.setVisibility(View.VISIBLE);
        }else{
            emptyTextView.setVisibility(View.INVISIBLE);
        }
        news = data;
        MainArrayAdapter adapter = new MainArrayAdapter(getBaseContext(),news);
        listView.setAdapter(adapter);
    }

    @Override
    public void onLoaderReset(@NonNull Loader<List<News>> loader) {

    }

}

File: Intermediate_food_ordering/app/src/main/java/com/example/intermediate_food_ordering/MainActivity.kt
package com.example.intermediate_food_ordering

import android.content.Intent
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.TextView
import androidx.annotation.RequiresApi
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.cart_counter.*
import kotlinx.android.synthetic.main.item_cart.*

class MainActivity : AppCompatActivity() {
    lateinit var cartCounterTextView: TextView //lateinit means we will initialize it later because we will go to onCreate, then inflate Options Menu.
    lateinit var adapter: MainArrayAdapter
    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.main, menu)
        cartCounterTextView = menu!!.findItem(R.id.main_shopping_cart_item).actionView.findViewById(R.id.cart_shopping_counter_textView) //we use actionView to get the custom layout
        adapter.setCartCounterTextView(cartCounterTextView)
        menu!!.getItem(0).actionView.setOnClickListener {
            intent = Intent(this,CartActivity::class.java)
            startActivity(intent)
        }
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return super.onOptionsItemSelected(item)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            setContentView(R.layout.activity_main)

        val menuItems: MutableList<com.example.intermediate_food_ordering.MenuItem> = ArrayList()

        menuItems.add(MenuItem("Apple","苹果", 3.69))
        menuItems.add(MenuItem("Fish","鱼", 6.99))
        menuItems.add(MenuItem("Beef","牛肉", 7.58))
        menuItems.add(MenuItem("Lamb","羊肉", 8.26))
        menuItems.add(MenuItem("Banana","香蕉", 4.78))
        menuItems.add(MenuItem("Water","水", 0.99))
        menuItems.add(MenuItem("Pear","梨", 2.45))
        menuItems.add(MenuItem("Milk","牛奶", 2.17))
        menuItems.add(MenuItem("Radish","萝卜", 3.78))
        menuItems.add(MenuItem("Cabbage","白菜", 2.57))

        adapter = MainArrayAdapter(this,menuItems)
        main_listView.adapter = adapter

    }

    override fun onStart() {
        super.onStart()
        if(this::cartCounterTextView.isInitialized) {  //check if initialize because we use lateinit
            cartCounterTextView.text = Main.getCartSize().toString()
            if(Main.addedItems.isEmpty()){
                cartCounterTextView.visibility = View.INVISIBLE
            }else{
                cartCounterTextView.visibility = View.VISIBLE
            }
        }
    }
}